### ISSUE TITLE
Unmounted screens renders on redux state update

### ISSUE THREAD

**Issue by jarrx** [wix/react-native-navigation#2618]
Is it expected behavior that screens that are not mounted or visible still gets updated and rerender when states in redux gets updated?

For example I have a screen, screen A, that uses a state from redux. When another screen, screen B, make changes to the same state as being used in screen A, screen A rerenders even though it's not visible.
I think this is strange since in screen A, componentWillUnmount is called when the screen closes (if shown as a modal).

Is there any solution to this? It's a problem since other parts of my app gets quite slow when the other screen that is not visible rerenders.

**Comment by pqkluan** [comment]
This is expected behavior from using Redux.

A change from the Redux store will update all the connected components via props.
When a props change, it will trigger the re-render cycle, updating the view node.

If your app is getting slow just by this, then it's mean the render method is doing too much work. You could fix this by implement `shouldComponentUpdate` method and reduce the amount of work the render method.

Post your code on StackOverflow for better support.

**Comment by andrewzey** [comment]
@pqkluan I don't agree that this behavior is at all expected or desirable, given the single-threaded nature of JavaScript. 

The app getting slow from re-rendering unmounted components doesn't necessarily mean that the render methods in those components are doing too much - as the calculation overhead is supposed to be at the selector layer. By design, selectors are the place for expensive calculations that can be memoized (as reselect does). Having all selectors for unmounted components potentially re-calculating, and then having those components possibly re-rendering is a terrible architectural decision in a single-threaded environment. 

It's disappointing that, apparently, every navigation library chooses to go down this route, as it makes it impossible to write performant apps without constantly resetting the stack. The problem with that, in the `react-native-navigation` library at this time, is that it prevents the ability to have nice slide screen transitions while also keeping a high-performance experience.

Possible solutions:

1. (Stop-gap) Implement `resetTo` to be able to do slide animations, so that a dev can reliably remove screens from the stack and prevent any unnecessary processing overhead stemming from them.

2. (React Native Navigation Architectural Fix): Implement JS worker multi-threading, with a communication bridge so that background screens can be rendered on a separate thread and not block the main thread. I realize this is a pretty tall order with respect to being able to share the same JS state, store, etc. - but this would be the "proper" solution to make sure screen transitions are as fast as possible, but without ever causing low priority calculations from blocking the main JS thread.

**Comment by pqkluan** [comment]
Hi @andrewzey, I don't think I fully understand what is the issue you want to address here.

`componentWillUnmount` only call when the Component is about to destroy following the concept of React lifecycle.

In RNN, there are only three scenarios for a screen component to be unmounted.
1. You pop a screen, that screen won't be accessible anymore so that it will be unmounted.
2. You dismiss a modal, every screen in that modal stack would be unmounted.
3. You start a new app; this will make all the screen in the old app to be unmounted.

Because of this, if your navigation stack has two screens that connect to the same part of the store, an update from that part should trigger the mapper functions in the connect an pass new props to these screens.

I said this is expected behavior from using Redux because if the pushed screen don't re-render when the props are changes, then when will it be updated, so the UI is in sync with the store data when user pop back?

That for the original question of this issue.

**And for your issue**

You could do many things to improve Redux performance:
* Don't dispatch unnecessary actions.
* Don't return a new state in the reducer if the data don't change.
* Use re-select to memorize heavy calculate selector.
* Use `makeMapStateToProps` to prepare re-use variable or function from the parent props.
* Only pass the necessary data to the Component in `mapStateToProps`.
* Implement `shouldComponentUpdate` depend only on changeable props.
* Break down the big container Component and divide them into smaller smart containers.

There are many things you could do to improve the app performance. If you have to resort to resetting the stack to keep the performance in check, **you are doing it wrong**. After a year of using RNN, I have never used that method in any of my apps. I do believe there are specific use-cases for that method, but it's not this one.

About multi-threading navigation...
Yeah, it sounds nice. But it's very tall order because this is an upstream issue up to JS nature. If there is a exists possible way to implement it, Node.js wouldn't be still running on a single thread right now.

**Comment by andrewzey** [comment]
@pqkluan Thanks for the reply!

I was a bit rushed when I wrote the above, so let me try to clarify what I was trying to get at.

One, I made an erroneous assumption about how RNN was working - I thought that
screens no longer visible in the stack were unmounted. As it turns out, that was entirely incorrect.

So, of course, that explains why those screens, which are implemented as connected components
(using `react-redux`'s `connect` HOC) are still re-rendering in the background, because they're still
subscribed to redux changes.

So as for the original question, you're 100% right. That's behavior as `designed` in RNN, because the
design of RNN does not unmount components in the stack that are not currently visible.

I contend that this poses a severe scaling limitation if you're dealing with a lot of connected components. Perhaps I can make it more clear why by running through your suggestions for improving Redux performance, and then provide a code sample of how we solved this problem:

> Don't dispatch unnecessary actions

:thumbsup:

> Don't return a new state in the reducer if the data don't change

:thumbsup:

> Use re-select to memorize heavy calculate selector.

Of course this is excellent advice, and we're using reselect extensively, but doesn't mitigate the problem I'm describing above. Because the connected component is still subscribed to Redux, it's `mapStateToProps` function will still execute, which means screens that are no longer visible may still be re-computing very expensive selectors. This is the primary cause of our particular performance problem, because we have some unavoidably expensive selectors, and as the screen stack increases, performance degrades quite quickly.

> Use `makeMapStateToProps` to prepare re-use variable or function from the parent props.

I'm actually not familiar with this, so I will take a look into it. Thank you! I found this medium article as a starting place https://medium.com/@parkerdan/react-reselect-and-redux-b34017f8194c.
Any other resources you could recommend?

> Only pass the necessary data to the Component in mapStateToProps.

:thumbsup:

> Implement shouldComponentUpdate depend only on changeable props.

Done, where necessary (where `react-redux`'s `connect` is insufficient), but doesn't mitigate the problem I described above, since `mapStateToProps` is invoked BEFORE `shouldComponentUpdate`.

> Break down the big container Component and divide them into smaller smart containers.

:thumbsup:

> If you have to resort to resetting the stack to keep the performance in check, you are doing it wrong.

I don't fully agree here. As I'll expand on below, I consider the fact that screens that are no longer visible are not unmounted to be a design flaw in the architecture of RNN. The performance degradation
of having even 5 or 6 screens in a stack is quite noticeable, and enough to result in dropped frames.

> Re: Multi-threading

Yes, that was just a fantasy solution on my part =).

My proposal regarding `resetTo` supporting properly-directioned slide transitions would be a stop-gap
for these types of performance problems, but I agree that they don't address the root cause of the
performance problem, which is that "backgrounded" screens in a stack are not unmounted.


So, here's what @awchang56 and I have done to workaround the performance problem in a VERY satisfactory way, which I believe can really help inform the design of a proper navigation solution
for react native. We've implemented this in our own NavigationHOC, but ideally it would be part of
library design:

1. Once a screen transition animation is complete, unmount the previous screen. This is accomplished by having a state variable in the NavigationHOC that tracks whether to render the WrappedComponent or not. Benefits of this are that anything in the render tree of the "background" screens (screens in the stack that are no longer visible) will no unsubscribe from redux - which means no expensive selectors, or any expensive operations in the component render paths will be run. This saves A LOT of resources

2. Wrap `pop` and `dismissModal` (the only ones we're using for popping off a stack) to check if the "background" screen we're going back to is now currently visible. If it is, trigger a render of the "background" screen, which will re-mount and render it.

The biggest potential downside is that re-drawing the "backgrounded" screen is slower in this paradigm
than how RNN currently does it. I don't think that's a problem at all, however, since if the component
is fast enough to draw when pushing, it's fast enough to draw when going back to it via popping.

Here's a simplified implementation of our `NavigationHOC` to show you what I mean:

```
import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {View} from 'react-native';

import {updateNavigationStack} from 'actions';
import {getNavigationStack} from 'selectors';
import {SCREEN_ANIMATION_DURATION} from 'constants';

const mapStateToProps = state => ({
  navigationStack: getNavigationStack(state),
});

export default function NavigationHOC(WrappedComponent, screenName) {
  class HOC extends React.Component {
    static propTypes = {
      dispatch: PropTypes.func,
      navigationStack: PropTypes.array, // Keeps track of screenNames in the navigationStack
      navigator: PropTypes.object,
    };

    static navigatorButtons = WrappedComponent.navigatorButtons;

    static navigatorStyle = {
      ...WrappedComponent.navigatorStyle,
      // Can set global navigatorStyles here
    };

    state = {
      shouldRender: true,
    };

    // used to prevent "double navigation" (eg. user quickly
    // taps on a button multiple times)
    _navigated = null;

    componentDidMount() {
      this.props.navigator.setOnNavigatorEvent(event => {
        // If we've navigated away from the screen, reset the _navigated
        // property used to prevent "double navigates" once we've come back
        if (event.id === 'didDisappear') {
          this._navigated = null;
        }

        // Override the native navigator methods
        const navigator = Object.create(this.props.navigator);
        navigator.resetTo = this.resetTo;
        navigator.push = this.pushTo;
        navigator.pop = this.pop;
        navigator.showModal = this.showModal;
        navigator.dismissModal = this.dismissModal;

        handleNavigatorEvents(event, navigator);
      });
    }

    componentWillReceiveProps(nextProps) {
      // If the screen is being navigated AWAY FROM, set the shouldRender to false, so that
      // the background component gets unmounted.
      // Except for Drawer, which needs to be rendered ADJACENT to the "previous" screen
      if (
        this.state.shouldRender &&
        !this.getIsScreenVisible(nextProps.navigationStack) &&
        screenName !== 'Drawer'
      ) {
        // NOTE: we want the screen we're hiding to still be visible until after the screen transition
        // animation is complete
        setTimeout(() => this.setState({shouldRender: false}), SCREEN_ANIMATION_DURATION);
      }
      // If the screen is being navigated TO, set the shouldRender to true, so that the
      // background component gets remounted, and re-rendered
      if (!this.state.shouldRender && this.getIsScreenVisible(nextProps.navigationStack)) {
        this.setState({shouldRender: true});
      }
    }

    getIsScreenVisible(navigationStack) {
      return navigationStack[navigationStack.length - 1] === screenName;
    }


    pushTo = screenOptions => {
      if (!this._navigated) {
        this.props.navigator.push(screenOptions);
        this.props.dispatch(updateNavigationStack('push', screenOptions.screen));
      }

      this._navigated = true;
    };

    pop = screenOptions => {
      if (!this._navigated) {
        this.props.navigator.pop(screenOptions);
        this.props.dispatch(updateNavigationStack('pop'));
      }

      this._navigated = true;
    };

    resetTo = screenOptions => {
      if (!this._navigated) {
        this.props.navigator.resetTo(screenOptions);
        this.props.dispatch(updateNavigationStack('resetTo', screenOptions.screen));
      }

      this._navigated = true;
    };

    showModal = screenOptions => {
      if (!this._navigated) {
        this.props.navigator.showModal(screenOptions);
        this.props.dispatch(updateNavigationStack('push', screenOptions.screen));
      }

      this._navigated = true;
    };

    dismissModal = screenOptions => {
      if (!this._navigated) {
        this.props.navigator.dismissModal(screenOptions);
        this.props.dispatch(updateNavigationStack('pop'));
      }

      this._navigated = true;
    };

    render() {
      return this.state.shouldRender ? (
        <WrappedComponent
          {...restProps}
          pushTo={this.pushTo}
          resetTo={this.resetTo}
          pop={this.pop}
          showModal={this.showModal}
          dismissModal={this.dismissModal}
        />
      ) : (
        // TODO: make backgroundColor the BG color of the wrapped components. This
        <View style={{backgroundColor: 'white', flex: 1}} />
      );
    }
  }
  return connect(mapStateToProps)(HOC);
}
```

Doing all this in a HOC that's rendered by `Navigation.registerComponent` and wraps every screen is
quite hacky, but this works very well, and has improved performance across the board dramatically,
without sacrificing anything in terms of user experience or code complexity outside of this HOC.

Once we're done with our current RN project, we plan on creating
our own proof of concept navigation library based on this paradigm, as both of the major react native
navigation libraries, `react-navigation` and `react-native-navigation` suffer from the same fatal flaw
of not un-mounting screens that are no longer visible in a stack. I firmly believe that it is a
substantial design mistake for a single-threaded environment. I agree that it won't be an issue for
most simple apps - but the lack of performant navigation in `react-native` definitely prevents it
from competing in the "big-leagues" for bigger, more complex applications.

**Comment by andrewzey** [comment]
There are numerous problems with what I wrote above, and I'll clarify them in a couple days when I have some time to write it up.

Long story short, I backed out everything I mentioned above, and the "proper" solution for us was memoizing `mapStateToProps` for backgrounded screens.

Thank you @pqkluan for suggesting `makeMapStateToProps` - that led me down the successful path!

**Comment by bneigher** [comment]
@andrewzey i just read this and am thinking about wrapping my screens with this to increase performance. In my case, there are no drops in FPS (maybe because hardware has evolved in 2 years), but instead my issue is a hot processor. Within 30seconds of my app being open (without even doing anything), the phone temperature gets noticing hotter. The only telemetry I can see as the suspect is a very high memory use, and a slight bit of a memory leak, but I have a feeling it’s due to having a lot of react internal listeners due to connect with redux store.  

I have already implemented a currentState mechanism so I can return false on shouldComponentUpdate, but the idea of not even having the other screens be connected to redux sounds like the real prize.

If you could summarize your conclusion I, and others would be stoked :)

**Comment by andrewzey** [comment]
@bneigher The mapStateToProps optimization I mentioned ended up being way too brittle.

Our team actually moved away from both react-navigation and react-native-navigation about a year ago to a purely in-house solution. It’s a pure JS solution, but without the enormous bloat and design issues of react-navigation. Ours implements a “rendered stack depth” of 4 screens. We maintain the stack history, and automatically unmount screens deeper than 4.

For any serious project where performance and minimizing bugs is important - I would strongly recommend writing your own minimal pure JS navigation system. I would regard react-native-navigation and react-navigation as good only for hobby projects and rapid MVPs (at least the version available as of a year ago). 

It took us less time to implement and switch to our own navigation system than to debug issues with the existing libraries. It dropped navigation bugs to zero, improved performance enormously, simplified the API, and reduced lines of code and depdencies 👍

**Comment by SrBrahma** [comment]
@andrewzey can you share your lib? I am making my own right now but I could use a ready one

**Comment by andrewzey** [comment]
@SrBrahma We might eventually consider open-sourcing it, but at the moment our solution is still fairly tied-in with our own code. Unfortunately it will take some considerable work to extract it out and think about what kind of interface to make for it, and that's not something our team can afford to dedicate time to doing right now =\.

Sometime next year we'll give some thought to giving back OSS to the community, but right now our resources are a bit strained.

Sorry!
